File: backend/app/utils/resume_parser.py
import re
from typing import List, Dict, Any

# Constants for feedback parsing
FEEDBACK_PATTERN = r'Original:\s*([^\n]+)\s*Grade:\s*([^\n]+)\s*Feedback:\s*([^\n]+(?:\n(?!- Option)[^\n]+)*)\s*(- Option 1:[^\n]+(?:\n(?!- Option)[^\n]+)*)\s*(- Option 2:[^\n]+(?:\n(?!Original:)[^\n]+)*)'

def parse_feedback_response(feedback_text: str) -> List[Dict[str, Any]]:
    """Parse the AI feedback response to extract structured feedback items."""
    
    if not feedback_text:
        return []
    
    matches = re.findall(FEEDBACK_PATTERN, feedback_text, re.DOTALL)
    items = []
    
    for match in matches:
        original_text = match[0].strip()
        grade = match[1].strip()
        feedback = match[2].strip()
        option1 = match[3].replace('- Option 1:', '').strip()
        option2 = match[4].replace('- Option 2:', '').strip()
        
        # Remove bullet points from original text
        original_text = re.sub(r'^[•\-]\s*', '', original_text)
        original_text = re.sub(r'^\s*[•\-]\s*', '', original_text)
        original_text = original_text.strip()
        
        # Skip items that are just intro text
        original_lower = original_text.lower()
        if ('detailed analysis' in original_lower or 
            'here\'s' in original_lower or
            'analysis of' in original_lower):
            continue
        
        # Clean up options
        complete_options = []
        for option in [option1, option2]:
            option_text = option.strip()
            if (len(option_text) > 15 and 
                not option_text.endswith('...') and 
                not option_text.endswith('..') and
                not option_text.endswith('etc') and
                not option_text.endswith('etc.')):
                complete_options.append(option_text)
        
        if len(complete_options) >= 2:
            items.append({
                "original": original_text,
                "grade": grade,
                "feedback": feedback,
                "options": complete_options
            })
    
    return items 

File: frontend/src/utils/resumeParser.js
// Constants for feedback parsing
const FEEDBACK_PATTERN = /Original:\s*([^\n]+)\s*Grade:\s*([^\n]+)\s*Feedback:\s*([^\n]+(?:\n(?!- Option)[^\n]+)*)\s*(- Option 1:[^\n]+(?:\n(?!- Option)[^\n]+)*)\s*(- Option 2:[^\n]+(?:\n(?!Original:)[^\n]+)*)/g;

/**
 * Parse raw feedback string into structured data
 * @param {string} raw - Raw feedback text from API
 * @returns {Array} Array of parsed feedback items
 */
export function parseFeedback(raw) {
  if (typeof raw !== "string") return [];
  
  const matches = [...raw.matchAll(FEEDBACK_PATTERN)];
  const pairs = [];
  
  for (const match of matches) {
    const originalText = match[1].trim().replace(/^[•\-]\s*/, '').replace(/^\s*[•\-]\s*/, '').trim();
    const grade = match[2].trim();
    const feedback = match[3].trim();
    const option1 = match[4].replace('- Option 1:', '').trim();
    const option2 = match[5].replace('- Option 2:', '').trim();
    
    // Skip items that are just intro text
    if (originalText.toLowerCase().includes('detailed analysis') || 
        originalText.toLowerCase().includes('here\'s') ||
        originalText.toLowerCase().includes('analysis of')) {
      continue;
    }
    
    // Clean up options
    const completeOptions = [];
    for (const option of [option1, option2]) {
      const optionText = option.trim();
      if (optionText.length > 15 && 
          !optionText.endsWith('...') && 
          !optionText.endsWith('..') &&
          !optionText.endsWith('etc') &&
          !optionText.endsWith('etc.')) {
        completeOptions.push(optionText);
      }
    }
    
    if (completeOptions.length >= 2) {
      pairs.push({
        original: originalText,
        grade: grade,
        feedback: feedback,
        options: completeOptions
      });
    }
  }
  
  console.log("Parsed pairs:", pairs);
  return pairs;
} 